database/gorm/query.go

```diff
@@ -701,6 +701,30 @@ func (r *QueryImpl) OrWhereIn(column string, values []any) ormcontract.Query {
 	return r.OrWhere(fmt.Sprintf("%s IN ?", column), values)
 }
 
+func (r *QueryImpl) WhereHas(table any, fk string, condition func(ormcontract.Query) (ormcontract.Query,error)) (ormcontract.Query,error) {
+
+	relatedModelQuery := NewQueryImplByInstance(r.instance.Model(table), r)
+
+	//wrap and modify query with the condition passed from function
+	modifiedQuery,_ := condition(relatedModelQuery)
+
+	// Extract the *gorm.DB instance from the modified QueryImpl
+	subQuery := modifiedQuery.(*QueryImpl).instance.Select(fk)
+
+	id := database.GetPrimaryField(table)
```

Comment:
There is a situation: The foreign key is not the primary id, User can customize it by tag, for example: 

```
type User struct {  
  gorm.Model  Name
  string  CompanyRefer int  
  Company  Company `gorm:"foreignKey:CompanyRefer"`
}

type Company struct {
  ID   int
  Name string
}
```

```
type User struct {
  gorm.Model  
  Name      string
  CompanyID string
  Company   Company `gorm:"references:Code"`
}

type Company struct {
  ID   int
  Code string
  Name string
}
```